drivers/usb/usbport/device.c

```diff
@@ -885,6 +885,75 @@ USBPORT_AbortTransfers(IN PDEVICE_OBJECT FdoDevice,
     }
 }
 
+PUSB2_TT_EXTENSION
+NTAPI
+USBPORT_GetTt(IN PDEVICE_OBJECT FdoDevice,
+              IN PUSBPORT_DEVICE_HANDLE HubDeviceHandle,
+              OUT PUSHORT OutPort,
+              OUT PUSBPORT_DEVICE_HANDLE * OutHubDeviceHandle)
+{
+    PUSBPORT_DEVICE_HANDLE DeviceHandle = HubDeviceHandle;
+    ULONG TtCount;
+    PLIST_ENTRY Entry;
+    PUSB2_TT_EXTENSION TtExtension = NULL;
+
+    DPRINT("USBPORT_GetTt: HubDeviceHandle - %p\n", HubDeviceHandle);
+
+    *OutHubDeviceHandle = NULL;
+
+    while (DeviceHandle->DeviceSpeed != UsbHighSpeed)
+    {
+        DPRINT("USBPORT_GetTt: DeviceHandle - %p, DeviceHandle->PortNumber - %X\n",
+               DeviceHandle,
+               DeviceHandle->PortNumber);
+
+        *OutPort = DeviceHandle->PortNumber;
+
+        DeviceHandle = DeviceHandle->HubDeviceHandle;
+
+        if (!DeviceHandle)
+            return NULL;
+    }
+
+    TtCount = DeviceHandle->TtCount;
+
+    if (!TtCount)
+        return NULL;
+
+    if (!IsListEmpty(&DeviceHandle->TtList))
+    {
+        Entry = DeviceHandle->TtList.Flink;
+
+        if (TtCount > 1)
+        {
+            while (Entry != &DeviceHandle->TtList)
+            {
+                ASSERT(Entry != NULL);
+
+                TtExtension = CONTAINING_RECORD(Entry,
+                                                USB2_TT_EXTENSION,
+                                                Link);
+
+                if (TtExtension->TtNumber == *OutPort)
+                    break;
+
+                Entry = Entry->Flink;
+                TtExtension = NULL;
+            }
+        }
+        else
+        {
+            TtExtension = CONTAINING_RECORD(Entry,
```

Comment:
So if there is only 1 item we do not care what OutPort is?
Should we assert on it?